File: projeto_ed_para_praticar/src/main/java/repository/dinamicas/lista/ListaDinamica.java
package repository.dinamicas.lista;

import repository.Listavel;
import repository.dinamicas.NoDuplo;
import repository.OverflowException;
import repository.UnderflowException;

public class ListaDinamica implements Listavel {
    private NoDuplo ponteiroInicio;
    private NoDuplo ponteiroFim;
    private int tamanhoMaximo;
    private int quantidade;

    public ListaDinamica(int tamanhoMaximo) {
        if (tamanhoMaximo <= 0) {
            throw new IllegalArgumentException("capacidade deve ser maior que zero.");
        }
        ponteiroInicio = null;
        ponteiroFim = null;
        this.tamanhoMaximo = tamanhoMaximo;
        quantidade = 0;
    }

    public ListaDinamica() {
        this(10);
    }

    @Override
    public void inserir(Object objeto, int posicao) {
        if (estaCheia()) {
            throw new OverflowException("Lista cheia");
        }
        if (posicao < 0 || posicao > quantidade) {
            throw new IndexOutOfBoundsException("Posição inválida");
        }

        NoDuplo ponteiroAuxiliar = ponteiroInicio;
        for(int i = 0; i < posicao; i++) {
            ponteiroAuxiliar = ponteiroAuxiliar.getProximo();
        }
        NoDuplo ant = (ponteiroAuxiliar != null) ? ponteiroAuxiliar.getAnterior() : ponteiroFim;
        NoDuplo prox = ponteiroAuxiliar;

        NoDuplo novoNo = new NoDuplo(objeto);
        novoNo.setProximo(prox);
        novoNo.setAnterior(ant);
        if (prox != null)
            prox.setAnterior(novoNo);
        else
            ponteiroFim = novoNo;


        if (ant != null)
            ant.setProximo(novoNo);
        else
            ponteiroInicio = novoNo;

        quantidade++;
    }

    @Override
    public void anexar(Object objeto) {
        if (estaCheia()) {
            throw new OverflowException("Lista cheia");
        }
        NoDuplo novoNo = new NoDuplo(objeto);
        novoNo.setAnterior(ponteiroFim);
        if (!estaVazia()) {
            ponteiroFim.setProximo(novoNo);
        } else {
            ponteiroInicio = novoNo;
        }
        ponteiroFim = novoNo;
        quantidade++;
    }

    @Override
    public Object selecionar(int posicao) {
        if (estaVazia()) {
            throw new UnderflowException("Lista vazia");
        }
        if (posicao < 0 || posicao >= quantidade) {
            throw new IndexOutOfBoundsException("Posição inválida");
        }
        NoDuplo ponteiroAuxiliar = ponteiroInicio;
        for (int i = 0; i < posicao; i++) {
            ponteiroAuxiliar = ponteiroAuxiliar.getProximo();
        }
        return ponteiroAuxiliar.getDado();
    }

    @Override
    public Object[] selecionarTodos() {
        Object[] arrayRetorno = new Object[quantidade];
        NoDuplo ponteiroAuxiliar = ponteiroInicio;
        for (int i = 0; i < quantidade; i++) {
            arrayRetorno[i] = ponteiroAuxiliar.getDado();
            ponteiroAuxiliar = ponteiroAuxiliar.getProximo();
        }
        return arrayRetorno;
    }

    @Override
    public void atualizar(Object objeto, int posicao) {
        if (estaVazia()) {
            throw new UnderflowException("Lista vazia");
        }
        if (posicao < 0 || posicao >= quantidade) {
            throw new IndexOutOfBoundsException("Posição inválida");
        }
        NoDuplo ponteiroAuxiliar = ponteiroInicio;
        for (int i = 0; i < posicao; i++) {
            ponteiroAuxiliar = ponteiroAuxiliar.getProximo();
        }
        ponteiroAuxiliar.setDado(objeto);
    }

    @Override
    public Object apagar(int posicao) {
        if (estaVazia()) {
            throw new UnderflowException("Lista vazia");
        }
        if (posicao < 0 || posicao >= quantidade) {
            throw new IndexOutOfBoundsException("Posição inválida");
        }
        NoDuplo ponteiroAuxiliar = ponteiroInicio;
        for (int i = 0; i < posicao; i++) {
            ponteiroAuxiliar = ponteiroAuxiliar.getProximo();
        }
        Object retorno = ponteiroAuxiliar.getDado();
        NoDuplo ant = ponteiroAuxiliar.getAnterior();
        NoDuplo prox = ponteiroAuxiliar.getProximo();
        if (ant != null) {
            ant.setProximo(prox);
        }else {
            ponteiroInicio = ponteiroInicio.getProximo();
        }
        if (prox != null) {
            prox.setAnterior(ant);
        } else {
            ponteiroFim = ponteiroFim.getAnterior();
        }
        quantidade--;
        return retorno;
    }

    @Override
    public void limpar() {
        ponteiroInicio = ponteiroFim = null;
        quantidade = 0;
    }

    @Override
    public int tamanho() {
        return quantidade;
    }

    @Override
    public boolean estaVazia() {
        return quantidade == 0;
    }

    @Override
    public boolean estaCheia() {
        return quantidade == tamanhoMaximo;
    }

    @Override
    public String imprimir() {
        String resultado = "";
        NoDuplo ponteiroAuxiliar = ponteiroInicio;
        for (int i = 0; i < quantidade; i++) {
            resultado += ponteiroAuxiliar.getDado();
            if (i != quantidade - 1) {
                resultado += ",";
            }
            ponteiroAuxiliar = ponteiroAuxiliar.getProximo();
        }
        return "[ " + resultado + " ]";
    }

    private NoDuplo noNaPosicao(int posicao) {
        NoDuplo p = ponteiroInicio;
        for (int i = 0; i < posicao; i++) {
            p = p.getProximo();
        }
        return p;
    }
}

File: lista/Lista Dinâmica/Kotlin/Lista/ListaDinamica.kt
class ListaDinamica(private val tamanho: Int = 10) : Listavel {

	private var ponteiroInicio: NoDuplo? = null
	private var ponteiroFim: NoDuplo? = null
	private var quantidade = 0

	//idêntico ao enfileirar de FilaDinamica
	override fun anexar(dado: Any?) {
		if (!estaCheia()) {
			val noTemp = NoDuplo(dado)
			//noTemp.dado = dado
			noTemp.anterior = ponteiroFim				
			if (!estaVazia())
				ponteiroFim?.proximo = noTemp
			else
				ponteiroInicio = noTemp
			
			ponteiroFim = noTemp
			quantidade = quantidade.inc()
		} else {
			println("Lista Cheia!")
		}
	}

	override fun selecionarTodos(): Array<Any?> {        
		var dadosAux: Array<Any?> = arrayOfNulls(quantidade)
		if (!estaVazia()) {
			var ponteiroAuxiliar = ponteiroInicio
			for (i in 0 until quantidade) {
				dadosAux[i] = ponteiroAuxiliar?.dado	
				ponteiroAuxiliar = ponteiroAuxiliar?.proximo
			}
		} else {
			println("Lista Vazia!")
		}
		return dadosAux
	}

	override fun selecionar(posicao: Int): Any? {
		var dadoAux: Any? = null
		if (!estaVazia()) {
			if (posicao >= 0 && posicao < quantidade) {
				////////////////////////////////
				//Codigo de posicionamento do ponteiro auxiliar, no nodo
				//que será feita alguma operação. Esse codigo é o mesmo
				//para os metodos update, delete, select e insert
				var ponteiroAuxiliar = ponteiroInicio
				for (i in 0 until posicao)
					ponteiroAuxiliar = ponteiroAuxiliar?.proximo
				///////////////////////////////
				dadoAux = ponteiroAuxiliar?.dado
			} else {
				println("Indice Inválido!")	
			}
		} else {
			println("Lista Vazia!")		
		}
		return dadoAux
	}

	override fun atualizar (posicao: Int, dado: Any?) {
		if (!estaVazia()) {
			if (posicao >= 0 && posicao < quantidade) {
				////////////////////////////////
				//Codigo de posicionamento do ponteiro auxiliar, no nodo
				//que será feita alguma operação. Esse codigo é o mesmo
				//para os metodos update, delete, select e insert
				var ponteiroAuxiliar = ponteiroInicio
				for (i in 0 until posicao)
					ponteiroAuxiliar = ponteiroAuxiliar?.proximo
				///////////////////////////////
				ponteiroAuxiliar?.dado = dado
			} else {
				println("Indice Inválido!")	
			}
		} else {
			println("Lista Vazia!")		
		}
	}

	override fun limpar() {
		ponteiroInicio = null
		ponteiroFim = null
		quantidade = 0
	}

	override fun apagarTodos(): Array<Any?> {
		var dadosAux: Array<Any?> = selecionarTodos()
		limpar()
		return dadosAux
	}

	override fun apagar(posicao: Int): Any? {
		var dadoAux: Any? = null
		if (!estaVazia()) {
			if (posicao >= 0 && posicao < quantidade) {
				////////////////////////////////
				//Codigo de posicionamento do ponteiro auxiliar, no nodo
				//que será feita alguma operação. Esse codigo é o mesmo
				//para os metodos update, delete, select e insert
				var ponteiroAuxiliar = ponteiroInicio
				for (i in 0 until posicao)
					ponteiroAuxiliar = ponteiroAuxiliar?.proximo
				///////////////////////////////
				dadoAux = ponteiroAuxiliar?.dado

				val ponteiroAnterior = ponteiroAuxiliar?.anterior
				val ponteiroProximo  = ponteiroAuxiliar?.proximo

				if (ponteiroAnterior != null) 
					ponteiroAnterior.proximo = ponteiroProximo
				//remocao do inicio, joga o ponteiro de inicio para o proximo nodo.
				else
					ponteiroInicio = ponteiroInicio?.proximo
				
				if (ponteiroProximo != null) 
					ponteiroProximo.anterior = ponteiroAnterior
				//remocao do fim, joga o ponteiro de fim para o nodo anterior.
				else
					ponteiroFim = ponteiroFim?.anterior

				quantidade = quantidade.dec()
			} else {
				println("Indice Inválido!")
			}
		} else {
			println("Lista Vazia!")
		}
		return dadoAux
	}

	override fun inserir(posicao: Int, dado: Any?) {
		if (!estaCheia()) {
			if (posicao >= 0 && posicao <= quantidade) {
				val noTemp = NoDuplo(dado)
				//noTemp.dado = dado
				
				////////////////////////////////
				//Codigo de posicionamento do ponteiro auxiliar, no nodo
				//que será feita alguma operação. Esse codigo é o mesmo
				//para os metodos update, delete, select e insert
				/*
				var ponteiroAuxiliar = ponteiroInicio
				for (i in 0 until posicao)
					ponteiroAuxiliar = ponteiroAuxiliar?.proximo
				///////////////////////////////
				val ponteiroProximo = ponteiroAuxiliar
				val ponteiroAnterior = ponteiroFim
				if (ponteiroAuxiliar != null)
					ponteiroAnterior = ponteiroAuxiliar.anterior
				*/
				var ponteiroAnterior: NoDuplo? = null
				var ponteiroProximo = ponteiroInicio

				for (i in 0 until posicao) {
					ponteiroAnterior = ponteiroProximo
					ponteiroProximo = ponteiroProximo?.proximo
				}

				//todas insercoes, exceto inicio
				if (ponteiroAnterior != null)						
					ponteiroAnterior.proximo = noTemp
				else	//insercao no inicio
					ponteiroInicio = noTemp

				//todas insercoes, exceto fim
				if (ponteiroProximo != null)
					ponteiroProximo.anterior = noTemp
				else	//insercao no fim
					ponteiroFim = noTemp						
			
				noTemp.proximo = ponteiroProximo
				noTemp.anterior = ponteiroAnterior						

				quantidade = quantidade.inc()
			} else {
				println("Indice Inválido!")
			}
		} else {
			println("Lista Cheia!")
		}
	}

	override fun estaCheia(): Boolean {
		return quantidade == tamanho
	}
	
	override fun estaVazia(): Boolean {
		return quantidade == 0
	}

	override fun tamanho(): Int {
		return quantidade
	}	

	override fun imprimir(): String {
		var ponteiroAuxiliar = ponteiroInicio
		var resultado = "["
		for (i in 0 until quantidade) {
			resultado += ponteiroAuxiliar?.dado
			if (i != quantidade-1)
				resultado += ","
			
			ponteiroAuxiliar = ponteiroAuxiliar?.proximo
		}
		return "$resultado]"
	}	
}